### PR TITLE
Table column text input

### PR DESCRIPTION
**Pull request by arjunpatel100**
[EVG-13205](https://jira.mongodb.org/browse/EVG-13205)

### Description 
Move text input filters to table columns. This PR is based on https://github.com/evergreen-ci/spruce/pull/867

### Screenshots
Screen Shot 2021-09-13 at 12 50 22 PM

### CODE REVIEW

File: cypress.json [hunk 1]
```diff
@@ -5,5 +5,8 @@
     "mochaFile": "bin/cypress/junit-[hash].xml"
   },
   "defaultCommandTimeout": 15000,
-  "retries": 4
+  "retries": {
+    "runMode": 4,
+    "openMode": 0
```
Comment (arjunpatel100) [inline cypress.json:10]: This prevents the test from retrying when ran locally

File: cypress/integration/task/test_table.ts [hunk 2]
```diff
@@ -78,14 +81,11 @@ describe("Tests Table", () => {
...
     });
   });
+
   describe("Test Status Selector", () => {
     beforeEach(() => {
       cy.visit(TESTS_ROUTE);
-      cy.get("[data-cy=test-status-select] > .cy-treeselect-bar").click();
-    });
-
-    it("Status select says 'No filters selected' by default", () => {
```
Comment (arjunpatel100) [inline cypress/integration/task/test_table.ts:87]: This test is removed since the input dropdown is now an icon embedded in the table column

File: cypress/integration/version/downstream_tasks.ts [hunk 3]
```diff
@@ -14,15 +14,17 @@ describe("Downstream Tasks Tab", () => {
     cy.visit(DOWNSTREAM_TASKS_ROUTE);
 
     cy.dataCy("project-accordion").should("have.length", 3);
-
+    cy.wait(0);
```
Comment (arjunpatel100) [inline cypress/integration/version/downstream_tasks.ts:17]: Added this line as a recommendation from https://github.com/cypress-io/cypress/issues/7306
`project-accordion` on line 18 was detaching from the between the time .first() and .click() were executed.

File: cypress/integration/version/task_filters.ts [hunk 4]
```diff
@@ -6,22 +6,21 @@ const patch = {
 };
 const path = `/version/${patch.id}`;
 const pathTasks = `${path}/tasks`;
-const pathURLWithFilters = `${pathTasks}?page=0&sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC&statuses=all,failed,success,dispatched,started,undispatched&taskName=test-thirdparty`;
+const pathURLWithFilters = `${pathTasks}?page=0&sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC&statuses=failed,success,dispatched,started,undispatched&taskName=test-thirdparty`;
```
Comment (arjunpatel100) [inline cypress/integration/version/task_filters.ts:9]: removed `all` because the statuses included in the URL don't match the `all` set

File: src/pages/hosts/HostsTable.tsx [hunk 5]
```diff
@@ -141,9 +142,10 @@ export const HostsTable: React.FC<Props> = ({
         placeholder: "Search ID",
         value: hostIdValue,
         onChange: onChangeHostId,
-        dataCy: "host-id-filter",
-        updateUrlParam: updateHostIdUrlParam,
-        resetUrlParam: resetHostIdUrlParam,
+        "data-cy": "host-id-filter",
+        onFilter: updateHostIdUrlParam,
+        onReset: resetHostIdUrlParam,
+        submitButtonCopy,
```
Comment (sophstad) [inline src/pages/hosts/HostsTable.tsx:148]: It doesn't look like `submitButtonCopy` is being used to override "Filter" 🤔 
image

**Review by sophstad — CHANGES_REQUESTED** [review]
This is sick! Great step toward standardization 🎉 

On the [antd demo](https://ant.design/components/table/#components-table-demo-head) I noticed the filter icon turns blue when a filter is applied. I wasn't able to produce this on Spruce and think it'd be nice to have!

File: cypress/integration/version/downstream_tasks.ts (hunk 3 above)
Comment (khelif96) [reply cypress/integration/version/downstream_tasks.ts:17]: Can we add a comment here with a description of why its necessary.

File: src/components/FilterInputControls/index.tsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    <ButtonWrapper>
+      <Button data-cy="reset-button" size="small" onClick={onClickReset}>
+        Reset
+      </Button>
+    </ButtonWrapper>
+    <Button
+      data-cy="filter-button"
+      size="small"
+      variant="primary"
+      onClick={onClickSubmit}
```
[the remark this answers is not in the document]
Comment (khelif96) [reply src/components/FilterInputControls/index.tsx:25]: ➕  Me too.

File: src/components/PatchTabs/downstreamTasks/reducer.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  | { type: "onResetTaskNameInput" }
+  | { type: "onResetVariantInput" }
+  | { type: "clearAllFilters" }
+  | { type: "onSort"; data: SortOrder[] }
+  | { type: "onChangePagination"; data: number }
+  | { type: "onChangeLimit"; data: number };
+
+export type State = QueryParamState & InputValueState;
+const page = { page: 0 };
+export const reducer = (state: State, action: Action) => {
```
Comment (khelif96) [inline src/components/PatchTabs/downstreamTasks/reducer.ts:40]: Great Refactor!

File: src/components/PatchTabs/downstreamTasks/reducer.ts [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+interface InputValueState {
+  baseStatusesInputVal: string[];
+  currentStatusesInputVal: string[];
+  taskNameInputVal: string;
+  variantInputVal: string;
+}
+
+export type Action =
+  | { type: "onChangeBaseStatusesSelector"; data: string[] }
```
Comment (khelif96) [inline src/components/PatchTabs/downstreamTasks/reducer.ts:21]: nit. instead of a generic `data` parameter we should prob include a more accurate name of the field that we are passing in. So the caller knows what kind of information should be passed in.

File: src/components/PatchTabs/downstreamTasks/reducer.ts [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+  | { type: "onResetStatusesSelector" }
+  | { type: "onResetTaskNameInput" }
+  | { type: "onResetVariantInput" }
+  | { type: "clearAllFilters" }
+  | { type: "onSort"; data: SortOrder[] }
+  | { type: "onChangePagination"; data: number }
+  | { type: "onChangeLimit"; data: number };
+
+export type State = QueryParamState & InputValueState;
+const page = { page: 0 };
```
Comment (khelif96) [inline src/components/PatchTabs/downstreamTasks/reducer.ts:39]: nit. resetPage?

File: src/hooks/useFilterInputChangeHandler.ts [hunk 10]
```diff
@@ -1,50 +1,70 @@
...
+  urlParam,
+  resetPage,
+  sendAnalyticsEvent = () => undefined,
+}: FilterHookParams): FilterHookResult<string> => {
+  const { search } = useLocation();
+  const updateQueryParams = useUpdateURLQueryParams();
+  const updateQueryParamWithDebounce = useMemo(
+    () => debounce(updateQueryParams, 250),
+    [updateQueryParams]
+  );
```
Comment (khelif96) [inline src/hooks/useFilterInputChangeHandler.ts:25]: Is this still necessary? Afaict we no longer update any values as we type. They only change when the user presses `filter` So a debounce seems unnecessary.

**Review by khelif96 — CHANGES_REQUESTED** [review]
Works Great! Just had a few minor comments but it looks almost done!

File: src/hooks/useFilterInputChangeHandler.ts (hunk 10 above)
Comment (arjunpatel100) [reply src/hooks/useFilterInputChangeHandler.ts:25]: this is used on the Patches page input where there is no submit btn

File: src/hooks/useUpdateURLQueryParams.ts [hunk 11]
```diff
@@ -6,14 +7,26 @@ const { stringifyQuery, parseQueryString } = queryString;
...
+      // remove any keys included in the delete list
+      const filteredParams = deleteKeys?.length
+        ? Object.entries(joinedParams)
+            .filter((p) => !deleteKeys.includes(p[0]))
+            .reduce((curr, accum) => ({ ...accum, [curr[0]]: curr[1] }), {})
+        : joinedParams;
+
+      replace(`${pathname}?${stringifyQuery(filteredParams)}`);
+    },
+    [replace, search, pathname]
```
Comment (khelif96) [inline src/hooks/useUpdateURLQueryParams.ts:26]: Can you explain why these changes are necessary? 
This is a pretty core piece of functionality in spruce. So i'm reluctant to make changes that change its behavior unless they are necessary.

File: src/pages/task/taskTabs/testsTable/getColumnsTemplate.tsx [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    sorter: true,
+    render: (status: string): JSX.Element => (
+      <span>
+        <Badge
+          variant={statusToBadgeColor[status] || Variant.LightGray}
+          key={status}
+        >
+          {statusCopy[status] || ""}
+        </Badge>
+      </span>
```
Comment (khelif96) [inline src/pages/task/taskTabs/testsTable/getColumnsTemplate.tsx:78]: same comments here. We could even turn this into a reusable component instead of rewriting it.

File: cypress/integration/job_logs.ts [hunk 13]
```diff
@@ -16,7 +16,7 @@ describe("Job Logs", () => {
       .should("have.attr", "href")
       .and(
         "equal",
-        `https://www.evergreen.mongodb.com/lobster/evergreen/complete-test/${taskId}/0/llama`
+        `https://evergreen.mongodb.com/lobster/evergreen/complete-test/${taskId}/0/llama`
```
[the remark this answers is not in the document]
Comment (arjunpatel100) [reply cypress/integration/job_logs.ts]: This is part of a past commit and isn't included in the current diff

File: src/components/FilterInputControls/index.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (arjunpatel100) [reply src/components/FilterInputControls/index.tsx:25]: taken care of here: https://github.com/evergreen-ci/spruce/pull/871/commits/64687a279e93070ce5b088f93178d3ca05609b61

File: cypress/integration/version/task_filters.ts (hunk 4 above)
Comment (arjunpatel100) [reply cypress/integration/version/task_filters.ts:9]: No, the task filter statuses change in EVG-15347

File: src/components/FilterBadges/index.tsx [hunk 14]
```diff
@@ -18,6 +18,9 @@ const hiddenQueryParams = new Set([
   ChartToggleQueryParams.chartType,
   PatchTasksQueryParams.Statuses,
   MainlineCommitQueryParams.SkipOrderNumber,
+  RequiredQueryParams.Sort,
+  RequiredQueryParams.Limit,
+  RequiredQueryParams.Page,
```
[the remark this answers is not in the document]
Comment (arjunpatel100) [reply src/components/FilterBadges/index.tsx]: You're right, these are unnecessary. The useStatusSelector hook was resetting page to 0 when a new status was pushed to the URL. I adjusted that in this commit:  https://github.com/evergreen-ci/spruce/pull/871/commits/29189a67e55725489cfca78ba72ce0bda0eb831f

File: src/hooks/useUpdateURLQueryParams.ts (hunk 11 above)
Comment (arjunpatel100) [reply src/hooks/useUpdateURLQueryParams.ts]: These changes allow the caller to indicate whether a query param should be remove from the URL completely. Instead we could set the query param value to undefined and achieve the same result.

File: src/components/Table/Filters.tsx [hunk 15]
```diff
@@ -196,3 +185,15 @@ const ButtonsWrapper = styled.div`
 const ButtonWrapper = styled.div`
   margin-right: 8px;
 `;
+
+const { blue } = uiColors;
```
Comment (khelif96) [inline src/components/Table/Filters.tsx:189]: nit. Can we destructure this at the top of the file for consistency.

**Review by khelif96 — APPROVED** [review]
LGTM Thanks for doing this!

File: src/components/Table/Filters.tsx [hunk 16]
```diff
@@ -155,3 +186,14 @@ const ButtonsWrapper = styled.div`
 const ButtonWrapper = styled.div`
   margin-right: 8px;
 `;
+
+interface StyledOutlinedProps {
+  active?: boolean;
+}
+const StyledFilterOutlined = styled(FilterOutlined)<StyledOutlinedProps>`
```
Comment (sophstad) [inline src/components/Table/Filters.tsx:193]: I wouldn't be opposed to using `FilterFilled` instead here, but either works!

**Review by sophstad — CHANGES_REQUESTED** [review]
Clicking "Clear all filters" correctly clears the filters and resets table contents, but does not update the column icon or dropdown

https://user-images.githubusercontent.com/9298431/134374513-1d237373-fc15-455b-8e1d-dcc17792f11c.mov

File: src/components/Table/Filters.tsx (hunk 16 above)
Comment (arjunpatel100) [reply src/components/Table/Filters.tsx:193]: I was unable to find a corresponding SearchFilled from the antd lib so I am going to keep as is.

**Review by sophstad — APPROVED** [review]
💯
